dcos/util.py

```diff
@@ -98,6 +100,63 @@ def remove_path_on_error(path):
         raise
 
 
+@contextlib.contextmanager
+def silent_outputs():
+    """A context manager for suppressing stdout / stderr, it sets their file
+    descriptors to os.devnull and then restores them.
```

Comment:
I am kind of afraid about the redirection to /dev/null. I mean, that precisely means that we throw away data. And here we do so rather thoroughly. I think even in the rather rare scenario that the CPython interpreter crashes with a Segmentation fault then even _that_ would be dumped into /dev/null. Mhhhhm... at least I'd appreciate some discussion around that!

Can't we build a context manager that buffers and that emits upon leaving the context?

Reply:
Even though we buffer this data, in case of a hard failure this wouldn't be printed either I guess. Also in my specific case, it is already prompting what it is going to try : 
```
If your browser didn't open, please go to the following link : (url)
```
Then it successfully opens a tab, but also prints warning messages. I don't think they should be printed back to the user, even after a delay, this can be confusing as things actually worked.

Reply:
@jgehrcke If an exception is thrown, the output of that will be printed as always (because the finally block restores the original stdout/stderr, which is back in place by the time the raised exception gets printed.

In the case of a segmentation fault, I don't see how buffering would help. A hard crash is a hard crash and the program would have no opportunity to restore the context and print the output anyway.

The only way to handle that would be to handle the OS signal from one of these extraordinary events and restore the IO redirection from there. I'm not sure it is worth the effort for such a large corner case.

Reply:
> In the case of a segmentation fault, I don't see how buffering would help. A hard crash is a hard crash and the program would have no opportunity to restore the context and print the output anyway.

That is correct. With the following statement:

> I think even in the rather rare scenario that the CPython interpreter crashes with a Segmentation fault then even that would be dumped into /dev/null

I tried to say that me maybe should just bend Python's I/O system within the context, and not completely bend the underlying file descriptors. In that case we'd still see the bytes directly emitted to the standard streams. When I thought about this I was assuming that the annoying warnings in questions are in fact emitted through Python's I/O system. I looked into the issue again and now see that that assumption is probably wrong: what we want to suppress is probably _not_ emitted through Python's I/O system, but by code that directly interacts with the OS through the OS file descriptors.

Is that correct?

That is, did we confirm that "just" manipulating `sys.stdout` and `sys.stderr` is not sufficient for suppressing the messages in question?

Thinking about this on that level I suppose a cleaner way to deal with this entire issue is to spawn a lean an clean child process via Python's `subprocess()` module. That process' sole purpose would be to use the `webbrowser` module. The trick would be to make it _not_ inherit the parent's standard streams, but to instead point its standard streams to pipes connected to the parent. And to then read its stdout and stderr streams asynchronously in the parent from a thread while prompting for the auth token. _After_ the prompt returns, the CLI process holds the complete data emitted by the browser-controlling process.

Just an idea for what I think would be a more solid architecture. I find it rather intrusive to redirect all stdout and stderr to /dev/null. If that _ever_ swallows helpful debugging information we will regret that.

Reply:
@jgehrcke : do you mean spawning a process that would invoke the python executable with inline code? If so I don't think this is an option, as we're shipping the dcos-cli as a self-contained binary which doesn't require a python executable to be present.

By cleaner do you mean that this would lead to less code? Because I find the current approach to be cleaner in some other sense, as it doesn't need to go through the additional cost of spawning another process in-between.

Reply:
> By cleaner do you mean that this would lead to less code?

No, it's all about

> I find it rather intrusive to redirect all stdout and stderr to /dev/null. If that ever swallows helpful debugging information we will regret that.

I was trying to suggest a way to that helps us retaining the data.